storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -2948,42 +2979,14 @@ fseg_free_step_not_header(
 		fil_block_check_type(*iblock, FIL_PAGE_INODE, mtr);
 	}
 
-	dberr_t err;
-	if (xdes_t* descr = fseg_get_first_extent(inode, space, mtr, &err)) {
-		/* Free the extent held by the segment */
-		return fseg_free_extent(inode, iblock, space,
-					xdes_get_offset(descr),
-					mtr
+	if (fseg_free_step_low(iblock, inode, space, mtr, header, false
 #ifdef BTR_CUR_HASH_ADAPT
-					, ahi
+	                       , ahi
 #endif /* BTR_CUR_HASH_ADAPT */
-					) != DB_SUCCESS;
-	} else if (err != DB_SUCCESS) {
-		return true;
-	}
-
-	/* Free a frag page */
-
-	ulint n = fseg_find_last_used_frag_page_slot(inode);
-
-	if (UNIV_UNLIKELY(n == ULINT_UNDEFINED)) {
-		return true;
-	}
-
-	uint32_t page_no = fseg_get_nth_frag_page_no(inode, n);
-
-	if (page_no == page_get_page_no(page_align(header))) {
+                               ) != DB_SUCCESS_LOCKED_REC) {
 		return true;
 	}
```

Comment:
We could replace `if (fseg_free_step_low(…) != DB_SUCCESS_LOCKED_REC)` with
```c++
	return fseg_free_step_low(…) != DB_SUCCESS_LOCKED_REC;
```
But, is that what we really want here? Wouldn’t we want `== DB_SUCCESS` instead? That is, our caller should keep freeing as long there was no corruption and we have not freed everything yet? Our callers are `btr_free_but_not_root()`, which is followed by `btr_free_root()`, and `trx_purge_free_segment()`. I’d suggest to carefully review the logic in both. A source code comment could be useful here.

On a related note, I see that `fseg_inode_free()` is checking for `DB_SUCCESS_LOCKED_REC`. It seems to be the right course of action there.

Reply:
If the `fseg_free_step_low()` returns any error code other than `DB_SUCCESS_LOCKED_REC`  then InnoDB
faced corruption or freed the segment successfully. So I think this condition is correct.